src/core/qgsvectorfilewriter.cpp

```diff
@@ -884,6 +907,55 @@ void QgsVectorFileWriter::init( QString vectorFileName,
             OGR_Fld_SetUnique( fld.get(), true );
           }
         }
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,5,0)
+        if ( mSetFieldDomains && srcDatabaseProviderConn )
+        {
+          const QString domainName = attrField.constraints().domainName();
+          if ( !domainName.isEmpty() )
+          {
+            bool canSetFieldDomainName = false;
+            if ( existingDestDomainNames.contains( domainName ) )
+            {
+              // If the target dataset already knows this field domain,
+              // we can directly assign its name to the new field.
+              canSetFieldDomainName = true;
+            }
+            else  if ( GDALDatasetTestCapability( mDS.get(), ODsCAddFieldDomain ) )
+            {
+              // Otherwise, if the output dataset can create field domains,
+              // - convert the QGIS field domain to a GDAL one
+              // - regiter it to the GDAL dataset
+              // - if successful, note that we know that field domain (if it
+              //   is shared by other fields)
+              // - assign its name to the new field.
+              std::unique_ptr<QgsFieldDomain> domain( srcDatabaseProviderConn->fieldDomain( domainName ) );
+              if ( domain )
+              {
+                OGRFieldDomainH hFieldDomain = QgsOgrUtils::convertFieldDomain( domain.get() );
+                if ( hFieldDomain )
+                {
+                  char *pszFailureReason = nullptr;
+                  if ( GDALDatasetAddFieldDomain( mDS.get(), hFieldDomain, &pszFailureReason ) )
+                  {
+                    existingDestDomainNames.insert( domainName );
+                    canSetFieldDomainName = true;
+                  }
+                  else
+                  {
+                    QgsDebugMsgLevel( QStringLiteral( "cannot create field domain: %1" ).arg( pszFailureReason ), 2 );
```

Comment:
Should we append this to mErrorMessage ?

Reply:
as far as I can see, mErrorMessage and mError are set only in the case of a fatal error. Here I was more in the mood that failure to create a field domain shouldn't prevent conversion of the layer. Otherwise we should have a GUI setting "Convert field domains" where the user could uncheck to bypass the error, but that's probably overkill.

Reply:
Right, lets leave for now then